### PR TITLE
Add function to factor tensor product of Pauli operators

### PR DESCRIPTION
**Pull request by marcusps**
@blakejohnson any comments? 

I'd also like to make `Pauli`s have a `[]` operator, but it wasn't clear to me if that was possible.  It would be nice to be able to pick out individual tensor factors without having to get all of them.

### CODE REVIEW

**Comment by blakejohnson** [comment]
I think you could define `getindex` on `Pauli`s to get that behavior.

File: src/Paulis.jl [hunk 1]
```diff
@@ -1,6 +1,6 @@
-import Base.complex
+import Base: complex, factor
 
-export Pauli, Id, X, Y, Z, allpaulis, paulieye, weight, complex, ∘
+export Pauli, Id, X, Y, Z, allpaulis, paulieye, weight, complex, ∘, factor
```
Comment (blakejohnson) [inline src/Paulis.jl]: The test failure seems to be related to exporting symbols also exported by Base. I *think*, but I am not sure, that one should not export a symbol already exported by Base.

Comment (blakejohnson) [inline src/Paulis.jl]: In any case, I don't think this PR introduces the test failures on Julia nightly, so I am fine with merging this as is.

**Comment by blakejohnson** [comment]
Looks like Julia 0.6 introduced using `∘` for function composition. So, we need to guard the export of that symbol on the julia version. We can fix that separately.

**Comment by marcusps** [comment]
OK, I'll merge and submit a separate PR for the \circ issue.

**Pull request merged**